{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "import numpy as np\n",
    "from scipy.sparse import csc_matrix, save_npz, load_npz\n",
    "import time\n",
    "import gzip\n",
    "from itertools import product\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Pull arguments\n",
    "reference = 'hg19'\n",
    "vcf_files = ['../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz', \n",
    "             '../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz',\n",
    "             '../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz']\n",
    "out_directory = '../split_gen_platinum'\n",
    "maxsize = 500000000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "194964483 Gens mismatch [1 0 1 1 1 1 1 1 1 1 1 1 1] [2 0 1 1 1 1 1 1 1 1 1 1 1]\n",
      "212227749 Gens mismatch [0 1 0 1 1 1 0 1 1 0 0 0 1] [0 1 0 1 1 0 0 1 1 0 0 0 1]\n",
      "2 382263\n",
      "(13, 382263)\n",
      "Completed in  18.174417734146118 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "55748731 Gens mismatch [0 1 1 0 0 1 1 0 1 1 0 0 1] [0 1 1 0 0 1 1 0 1 1 0 1 1]\n",
      "1 396795\n",
      "(13, 396795)\n",
      "Completed in  18.317864894866943 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "0 333341\n",
      "(13, 333341)\n",
      "Completed in  16.685539960861206 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "0 359420\n",
      "(13, 359420)\n",
      "Completed in  17.469725847244263 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "13976819 Gens mismatch [1 1 1 0 1 1 1 2 0 1 0 1 2] [1 1 1 0 1 1 1 1 0 1 0 1 1]\n",
      "106976722 Gens mismatch [1 0 1 1 0 1 0 1 0 1 0 1 1] [1 0 1 1 0 1 0 1 0 0 0 1 1]\n",
      "2 298892\n",
      "(13, 298892)\n",
      "Completed in  16.060200929641724 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "0 318274\n",
      "(13, 318274)\n",
      "Completed in  16.508931159973145 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "107824572 Gens mismatch [0 1 0 1 1 0 1 1 0 0 0 0 1] [0 1 0 1 1 1 1 1 1 0 1 0 1]\n",
      "1 281263\n",
      "(13, 281263)\n",
      "Completed in  15.630668878555298 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "0 255069\n",
      "(13, 255069)\n",
      "Completed in  14.831913948059082 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "0 219594\n",
      "(13, 219594)\n",
      "Completed in  13.963639974594116 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "0 243740\n",
      "(13, 243740)\n",
      "Completed in  14.456036806106567 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "0 245110\n",
      "(13, 245110)\n",
      "Completed in  14.652414321899414 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "0 229079\n",
      "(13, 229079)\n",
      "Completed in  14.05094575881958 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "0 183724\n",
      "(13, 183724)\n",
      "Completed in  12.960637092590332 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "0 151544\n",
      "(13, 151544)\n",
      "Completed in  11.97673511505127 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "0 143166\n",
      "(13, 143166)\n",
      "Completed in  12.13083004951477 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "0 154213\n",
      "(13, 154213)\n",
      "Completed in  12.23791790008545 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "51823076 Gens mismatch [0 2 1 1 1 1 1 1 1 1 1 1 2] [0 2 1 1 1 1 1 1 1 1 1 1 1]\n",
      "1 133842\n",
      "(13, 133842)\n",
      "Completed in  11.616289854049683 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "57257048 Gens mismatch [0 1 0 0 0 1 0 0 0 0 0 0 0] [1 1 1 1 1 1 1 1 1 1 1 1 1]\n",
      "76878390 Gens mismatch [0 1 1 0 1 0 0 1 0 0 0 0 1] [1 1 1 1 1 1 1 1 1 1 1 1 1]\n",
      "2 138663\n",
      "(13, 138663)\n",
      "Completed in  11.69355297088623 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "44530694 Gens mismatch [1 0 1 1 1 0 0 0 1 0 1 0 0] [1 1 1 1 1 1 1 1 1 1 1 1 1]\n",
      "1 110012\n",
      "(13, 110012)\n",
      "Completed in  11.047881841659546 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "6055696 Gens mismatch [1 0 1 1 0 1 0 1 0 1 0 1 0] [1 0 1 1 0 1 1 1 0 1 0 1 0]\n",
      "1 107449\n",
      "(13, 107449)\n",
      "Completed in  10.914196968078613 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "0 79415\n",
      "(13, 79415)\n",
      "Completed in  10.21255898475647 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "0 65006\n",
      "(13, 65006)\n",
      "Completed in  9.8523690700531 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "0 127699\n",
      "(13, 127699)\n",
      "Completed in  12.18364691734314 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n"
     ]
    },
    {
     "ename": "IndexError",
     "evalue": "too many indices for array",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-13-081dc1159524>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     77\u001b[0m     \u001b[0;31m# reorder since data is coming from multiple files\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     78\u001b[0m     \u001b[0mchrom_coord\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0masarray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mchrom_coord\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mint\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 79\u001b[0;31m     \u001b[0mordered_indices\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0margsort\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mchrom_coord\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     80\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     81\u001b[0m     \u001b[0mchrom_coord\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mchrom_coord\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mordered_indices\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m:\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mIndexError\u001b[0m: too many indices for array"
     ]
    }
   ],
   "source": [
    "for chrom in ['chr%d' % c for c in range(1, 23)] + ['chrX', 'chrY']:\n",
    "    chrom_int = 23 if chrom == 'chrX' else 24 if chrom == 'chrY' else 25 if chrom == 'chrMT' else int(chrom[3:])\n",
    "\n",
    "    t0 = time.time()\n",
    "\n",
    "    # pull sample_ids and write to file\n",
    "    with gzip.open(vcf_files[0], 'rt') as f:\n",
    "        # Skip header\n",
    "        line = next(f)\n",
    "        while line.startswith('##'):\n",
    "            line = next(f)\n",
    "\n",
    "    with open('%s/chr.%s.gen.samples.txt' % (out_directory, chrom[3:]), 'w+') as sample_f:\n",
    "        sample_ids = line.strip().split('\\t')[9:]\n",
    "        sample_f.write('\\n'.join(sample_ids))\n",
    "\n",
    "    sample_id_to_index = dict([(x, i) for i, x in enumerate(sample_ids)])\n",
    "    m = len(sample_ids)\n",
    "    print('Num individuals with genomic data', len(sample_ids))\n",
    "\n",
    "    # Pull data from vcf\n",
    "    variant_info = []\n",
    "    chrom_coord = []\n",
    "\n",
    "    # Pull genotypes from vcf\n",
    "    m = len(sample_ids)\n",
    "    data, indices, indptr = np.zeros((maxsize,), dtype=np.int8), np.zeros((maxsize,), dtype=int), [0]\n",
    "    gen_mapping = {'./.': -1, '0/0': 0, '0|0': 0, '0/1': 1, '0|1': 1, '1/0': 1, '1|0': 1, '1/1': 2, '1|1': 2}\n",
    "    index, num_lines = 0, 0\n",
    "    \n",
    "    refs = []\n",
    "    alts = []\n",
    "\n",
    "    for vcf_file in vcf_files:\n",
    "        print(vcf_file)\n",
    "        with gzip.open(vcf_file, 'rt') as f:\n",
    "\n",
    "            # Skip header\n",
    "            line = next(f)\n",
    "            while line.startswith('##'):\n",
    "                line = next(f)\n",
    "\n",
    "            sample_ids = line.strip().split('\\t')[9:]\n",
    "            col_indices = [sample_id_to_index[x] for x in sample_ids]\n",
    "\n",
    "            line = next(f)\n",
    "            for line in f:\n",
    "                pieces = line.split('\\t', maxsplit=1)\n",
    "\n",
    "                if pieces[0] == chrom:\n",
    "                    pieces = line.strip().split('\\t')\n",
    "                    _, pos, _, ref, alt = pieces[:5]\n",
    "                    is_biallelic_snp = len(ref) == 1 and len(alt) == 1 and ref != '.' and alt != '.'\n",
    "                    \n",
    "                    if is_biallelic_snp and 'bwa_gatk' in pieces[7]:\n",
    "                        refs.append(ref)\n",
    "                        alts.append(alt)\n",
    "                        fmt = pieces[8].strip().split(':')\n",
    "\n",
    "                        # Write variant to file\n",
    "                        variant_info.append('\\t'.join(pieces[:9]) + '\\n')\n",
    "                        chrom_coord.append((chrom_int, int(pos), is_biallelic_snp, True))\n",
    "\n",
    "                        # Pull out genotypes\n",
    "                        gen_index = fmt.index('GT')\n",
    "                        for i, piece in zip(col_indices, pieces[9:]):\n",
    "                            segment = piece.split(':', maxsplit=gen_index+1)\n",
    "                            gt = gen_mapping.get(segment[gen_index], -1) # For now we mark multi-base loci as unknown\n",
    "\n",
    "                            if gt != 0:\n",
    "                                indices[index] = i\n",
    "                                data[index] = gt\n",
    "                                index += 1\n",
    "                        indptr.append(index)\n",
    "                        num_lines += 1\n",
    "\n",
    "    # reorder since data is coming from multiple files\n",
    "    chrom_coord = np.asarray(chrom_coord, dtype=int)\n",
    "    ordered_indices = np.argsort(chrom_coord[:, 1])\n",
    "\n",
    "    chrom_coord = chrom_coord[ordered_indices, :]\n",
    "    gen = csc_matrix((data[:index], indices[:index], indptr), shape=(m, num_lines), dtype=np.int8)[:, ordered_indices]\n",
    "    refs = [refs[i] for i in ordered_indices]\n",
    "    alts = [alts[i] for i in ordered_indices]\n",
    "    \n",
    "    # remove sites with same alt and genotypes\n",
    "    keep_site = np.ones((chrom_coord.shape[0],), dtype=bool)\n",
    "    for i in np.where(chrom_coord[:-1, 1] == chrom_coord[1:, 1])[0]:\n",
    "        # positions are the same\n",
    "        if alts[i] == alts[i+1] and np.all(gen[:, i].A.flatten() == gen[:, i+1].A.flatten()):\n",
    "            keep_site[i+1] = False\n",
    "        elif alts[i] != alts[i+1]:\n",
    "            keep_site[i] = False\n",
    "            keep_site[i+1] = False\n",
    "            print(chrom_coord[i, 1], 'Alt mismatch', alts[i], alts[i+1], refs[i], refs[i+1])\n",
    "        else:\n",
    "            keep_site[i] = False\n",
    "            keep_site[i+1] = False\n",
    "            print(chrom_coord[i, 1], 'Gens mismatch', gen[:, i].A.flatten(), gen[:, i+1].A.flatten())\n",
    "    print(np.sum(chrom_coord[:-1, 1] == chrom_coord[1:, 1]), chrom_coord.shape[0])\n",
    "    \n",
    "    with gzip.open('%s/chr.%s.gen.variants.txt.gz' % (out_directory, chrom[3:]), 'wt') as variant_f:\n",
    "        for i in ordered_indices:\n",
    "            variant_f.write(variant_info[i])\n",
    "\n",
    "    # Save to file\n",
    "    print(gen.shape)\n",
    "    save_npz('%s/chr.%s.gen' % (out_directory, chrom[3:]), gen[:, keep_site])\n",
    "    np.save('%s/chr.%s.gen.coordinates' % (out_directory, chrom[3:]), chrom_coord[keep_site, :])\n",
    "\n",
    "    print('Completed in ', time.time()-t0, 'sec')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "chrom_coord = np.load('../split_gen_platinum/chr.1.gen.coordinates.npy')[:, 1]\n",
    "gen_data = load_npz('../split_gen_platinum/chr.1.gen.npz')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(13, 382259) (382259,)\n"
     ]
    }
   ],
   "source": [
    "print(gen_data.shape, chrom_coord.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(1, chrom_coord.shape[0]):\n",
    "    if chrom_coord[i] == chrom_coord[i-1]:\n",
    "        print(chrom_coord[i-1], chrom_coord[i])\n",
    "        print(gen_data[:, i-1:i+1].A)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "match_next = chrom_coord[1:] == chrom_coord[:-1]\n",
    "gen_match_next = "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
