{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "import numpy as np\n",
    "from scipy.sparse import csc_matrix, save_npz, load_npz\n",
    "import time\n",
    "import gzip\n",
    "from itertools import product\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Pull arguments\n",
    "reference = 'hg19'\n",
    "vcf_files = ['../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz', \n",
    "             '../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz',\n",
    "             '../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz']\n",
    "out_directory = '../split_gen_platinum'\n",
    "maxsize = 500000000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "(13, 408531)\n",
      "Completed in  19.66784381866455 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "(13, 423601)\n",
      "Completed in  20.548337936401367 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "(13, 352114)\n",
      "Completed in  18.94524884223938 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "(13, 379927)\n",
      "Completed in  19.441507816314697 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "(13, 316528)\n",
      "Completed in  19.060844898223877 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "(13, 337572)\n",
      "Completed in  19.4302659034729 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "(13, 307997)\n",
      "Completed in  16.670969247817993 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "(13, 271256)\n",
      "Completed in  15.932062864303589 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "(13, 241545)\n",
      "Completed in  15.787659168243408 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "(13, 263835)\n",
      "Completed in  17.40366291999817 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "(13, 263657)\n",
      "Completed in  17.388401985168457 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "(13, 247147)\n",
      "Completed in  16.230600833892822 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "(13, 193964)\n",
      "Completed in  14.539880990982056 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "(13, 161459)\n",
      "Completed in  14.263708114624023 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "(13, 156752)\n",
      "Completed in  13.446372747421265 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "(13, 168251)\n",
      "Completed in  14.675538063049316 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "(13, 146212)\n",
      "Completed in  13.370973110198975 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "(13, 147817)\n",
      "Completed in  12.914270162582397 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "(13, 123848)\n",
      "Completed in  12.582389116287231 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "(13, 116038)\n",
      "Completed in  12.313359022140503 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "(13, 85720)\n",
      "Completed in  11.335952043533325 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "(13, 73450)\n",
      "Completed in  11.403707027435303 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n",
      "(13, 132337)\n",
      "Completed in  12.745614051818848 sec\n",
      "Num individuals with genomic data 13\n",
      "../data/GenotypeFiles/High_Confidence_Calls_HG19.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_filtered_ped_consistent.vcf.gz\n",
      "../data/GenotypeFiles/manuscript_supplement_hq_fails.vcf.gz\n"
     ]
    },
    {
     "ename": "IndexError",
     "evalue": "too many indices for array",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-36-b37b100d808c>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     72\u001b[0m     \u001b[0;31m# reorder since data is coming from multiple files\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     73\u001b[0m     \u001b[0mchrom_coord\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0masarray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mchrom_coord\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mint\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 74\u001b[0;31m     \u001b[0mordered_indices\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0margsort\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mchrom_coord\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     75\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     76\u001b[0m     \u001b[0mchrom_coord\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mchrom_coord\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mordered_indices\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m:\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mIndexError\u001b[0m: too many indices for array"
     ]
    }
   ],
   "source": [
    "for chrom in ['chr%d' % c for c in range(1, 23)] + ['chrX', 'chrY']:\n",
    "    chrom_int = 23 if chrom == 'chrX' else 24 if chrom == 'chrY' else 25 if chrom == 'chrMT' else int(chrom[3:])\n",
    "\n",
    "    t0 = time.time()\n",
    "\n",
    "    # pull sample_ids and write to file\n",
    "    with gzip.open(vcf_files[0], 'rt') as f:\n",
    "        # Skip header\n",
    "        line = next(f)\n",
    "        while line.startswith('##'):\n",
    "            line = next(f)\n",
    "\n",
    "    with open('%s/chr.%s.gen.samples.txt' % (out_directory, chrom[3:]), 'w+') as sample_f:\n",
    "        sample_ids = line.strip().split('\\t')[9:]\n",
    "        sample_f.write('\\n'.join(sample_ids))\n",
    "\n",
    "    sample_id_to_index = dict([(x, i) for i, x in enumerate(sample_ids)])\n",
    "    m = len(sample_ids)\n",
    "    print('Num individuals with genomic data', len(sample_ids))\n",
    "\n",
    "    # Pull data from vcf\n",
    "    variant_info = []\n",
    "    chrom_coord = []\n",
    "\n",
    "    # Pull genotypes from vcf\n",
    "    m = len(sample_ids)\n",
    "    data, indices, indptr = np.zeros((maxsize,), dtype=np.int8), np.zeros((maxsize,), dtype=int), [0]\n",
    "    gen_mapping = {'./.': -1, '0/0': 0, '0|0': 0, '0/1': 1, '0|1': 1, '1/0': 1, '1|0': 1, '1/1': 2, '1|1': 2}\n",
    "    index, num_lines = 0, 0\n",
    "\n",
    "    for vcf_file in vcf_files:\n",
    "        print(vcf_file)\n",
    "        with gzip.open(vcf_file, 'rt') as f:\n",
    "\n",
    "            # Skip header\n",
    "            line = next(f)\n",
    "            while line.startswith('##'):\n",
    "                line = next(f)\n",
    "\n",
    "            sample_ids = line.strip().split('\\t')[9:]\n",
    "            col_indices = [sample_id_to_index[x] for x in sample_ids]\n",
    "\n",
    "            line = next(f)\n",
    "            for line in f:\n",
    "                pieces = line.split('\\t', maxsplit=1)\n",
    "\n",
    "                if pieces[0] == chrom:\n",
    "                    pieces = line.strip().split('\\t')\n",
    "                    _, pos, _, ref, alt = pieces[:5]\n",
    "                    is_biallelic_snp = len(ref) == 1 and len(alt) == 1 and ref != '.' and alt != '.'\n",
    "                    \n",
    "                    if is_biallelic_snp:\n",
    "                        fmt = pieces[8].strip().split(':')\n",
    "\n",
    "                        # Write variant to file\n",
    "                        variant_info.append('\\t'.join(pieces[:9]) + '\\n')\n",
    "                        chrom_coord.append((chrom_int, int(pos), is_biallelic_snp, True))\n",
    "\n",
    "                        # Pull out genotypes\n",
    "                        gen_index = fmt.index('GT')\n",
    "                        for i, piece in zip(col_indices, pieces[9:]):\n",
    "                            segment = piece.split(':', maxsplit=gen_index+1)\n",
    "                            gt = gen_mapping.get(segment[gen_index], -1) # For now we mark multi-base loci as unknown\n",
    "\n",
    "                            if gt != 0:\n",
    "                                indices[index] = i\n",
    "                                data[index] = gt\n",
    "                                index += 1\n",
    "                        indptr.append(index)\n",
    "                        num_lines += 1\n",
    "\n",
    "    # reorder since data is coming from multiple files\n",
    "    chrom_coord = np.asarray(chrom_coord, dtype=int)\n",
    "    ordered_indices = np.argsort(chrom_coord[:, 1])\n",
    "\n",
    "    chrom_coord = chrom_coord[ordered_indices, :]\n",
    "    gen = csc_matrix((data[:index], indices[:index], indptr), shape=(m, num_lines), dtype=np.int8)[:, ordered_indices]\n",
    "\n",
    "    with gzip.open('%s/chr.%s.gen.variants.txt.gz' % (out_directory, chrom[3:]), 'wt') as variant_f:\n",
    "        for i in ordered_indices:\n",
    "            variant_f.write(variant_info[i])\n",
    "\n",
    "    # Save to file\n",
    "    print(gen.shape)\n",
    "    save_npz('%s/chr.%s.gen' % (out_directory, chrom[3:]), gen)\n",
    "    np.save('%s/chr.%s.gen.coordinates' % (out_directory, chrom[3:]), chrom_coord)\n",
    "\n",
    "    print('Completed in ', time.time()-t0, 'sec')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "chrom_coord = np.load('../split_gen_platinum/chr.17.gen.coordinates.npy')[:, 1]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[     828      834     1389 ... 81149900 81151539 81151920]\n"
     ]
    }
   ],
   "source": [
    "print(chrom_coord)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(1, chrom_coord.shape[0]):\n",
    "    if chrom_coord[i] < chrom_coord[i-1]:\n",
    "        print(chrom_coord[i-1], chrom_coord[i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "match_next = chrom_coord[1:] == chrom_coord[:-1]\n",
    "gen_match_next = "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
