{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prep data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "import time\n",
    "from os import listdir\n",
    "import gzip\n",
    "\n",
    "from collections import Counter, Hashable, defaultdict\n",
    "from itertools import chain, product\n",
    "\n",
    "import numpy as np\n",
    "from scipy import sparse\n",
    "import random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# From GRCh37.p13 https://www.ncbi.nlm.nih.gov/grc/human/data?asm=GRCh37.p13\n",
    "chrom_lengths = {\n",
    "\t'1': 249250621,\n",
    "\t'2': 243199373,\n",
    "\t'3': 198022430,\n",
    "\t'4': 191154276,\n",
    "\t'5': 180915260,\n",
    "\t'6': 171115067,\n",
    "\t'7': 159138663,\n",
    "\t'8': 146364022,\n",
    "\t'9': 141213431,\n",
    "\t'10': 135534747,\n",
    "\t'11': 135006516,\n",
    "\t'12': 133851895,\n",
    "\t'13': 115169878,\n",
    "\t'14': 107349540,\n",
    "\t'15': 102531392,\n",
    "\t'16': 90354753,\n",
    "\t'17': 81195210,\n",
    "\t'18': 78077248,\n",
    "\t'19': 59128983,\n",
    "\t'20': 63025520,\n",
    "\t'21': 48129895,\n",
    "\t'22': 51304566,\n",
    "\t'X': 155270560,\n",
    "\t'Y': 59373566\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "chrom = '22'\n",
    "m = 4\n",
    "ped_file = '../data/160826.ped'\n",
    "data_dir = '../split_gen_miss'\n",
    "# out_dir = sys.argv[5]\n",
    "batch_size = None\n",
    "batch_num = None\n",
    "batch_offset = None\n",
    "\n",
    "error_rate = .02\n",
    "smooth = 5000\n",
    "chrom_length = chrom_lengths[chrom]\n",
    "\n",
    "# genotype (pred, obs): cost\n",
    "g_cost = {\n",
    "\t(-1, -1): 0,\n",
    "\t(-1, 0): 1,\n",
    "\t(-1, 1): 1,\n",
    "\t(-1, 2): 1,\n",
    "\t(0, -1): 0,\n",
    "\t(0, 0): 0,\n",
    "\t(0, 1): 1,\n",
    "\t(0, 2): 2,\n",
    "\t(1, -1): 0,\n",
    "\t(1, 0): 1,\n",
    "\t(1, 1): 0,\n",
    "\t(1, 2): 1,\n",
    "\t(2, -1): 0,\n",
    "\t(2, 0): 2,\n",
    "\t(2, 1): 1,\n",
    "\t(2, 2): 0\n",
    "}\n",
    "\n",
    "shift_costs = [10]*4 + [500]*(2*(m-2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set up filenames\n",
    "sample_file = '%s/chr.%s.gen.samples.txt' % (data_dir, chrom)\n",
    "coord_file = '%s/chr.%s.gen.coordinates.npy' % (data_dir,  chrom)\n",
    "gen_files = sorted([f for f in listdir(data_dir) if ('chr.%s' % chrom) in f and 'gen.npz' in f])\n",
    "\n",
    "#fam_output_file = '%s/chr.%s.familysize.%s.families.txt' % (out_dir, chrom, m)\n",
    "#phase_output_file = '%s/chr.%s.familysize.%s.phased.txt' % (out_dir, chrom, m)\n",
    "\n",
    "# adapt to batch_size\n",
    "if batch_size is not None:\n",
    "\tbatch_offset = batch_size*batch_num\n",
    "\t#fam_output_file = fam_output_file[:-4] + str(batch_num) + '.txt'\n",
    "\t#phase_output_file = phase_output_file[:-4] + str(batch_num) + '.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "families with sequence data 886\n",
      "families of size 4: 436\n",
      "families pulled 4: 436\n"
     ]
    }
   ],
   "source": [
    "def pull_families(sample_file, batch_size=None, batch_offset=None):\n",
    "\t# pull families with sequence data\n",
    "\twith open(sample_file, 'r') as f:\n",
    "\t\tsample_ids = [line.strip() for line in f]\n",
    "\n",
    "\t# pull families from ped file\n",
    "\tfamilies = dict()\n",
    "\twith open(ped_file, 'r') as f:\t\n",
    "\t    for line in f:\n",
    "\t        pieces = line.strip().split('\\t')\n",
    "\t        fam_id, child_id, f_id, m_id = pieces[0:4]\n",
    "\n",
    "\t        if child_id in sample_ids and f_id in sample_ids and m_id in sample_ids:\n",
    "\t        \tif (fam_id, m_id, f_id) not in families:\n",
    "\t        \t\tfamilies[(fam_id, m_id, f_id)] = [m_id, f_id]\n",
    "\t        \tfamilies[(fam_id, m_id, f_id)].append(child_id)\n",
    "\n",
    "\t# randomly permute children\n",
    "\tfamilies = dict([(k, x[:2]+random.sample(x[2:], len(x)-2)) for k, x in families.items()])\n",
    "\tprint('families with sequence data', len(families))\n",
    "\n",
    "\tfamilies_of_this_size = [(fkey, inds) for fkey, inds in families.items() if len(inds) == m]\n",
    "\tprint('families of size %d: %d' % (m, len(families_of_this_size)))\n",
    "\n",
    "\t# limit to batch\n",
    "\tif batch_size is not None:\n",
    "\t\tfamily_keys = set(sorted([x[0] for x in families_of_this_size])[batch_offset:(batch_size+batch_offset)])\n",
    "\t\tfamilies_of_this_size = [(k, v) for k, v in families_of_this_size if k in family_keys]\n",
    "\t\n",
    "\tprint('families pulled %d: %d' % (m, len(families_of_this_size)))\n",
    "\treturn families_of_this_size\n",
    "\n",
    "families_of_this_size = pull_families(sample_file, batch_size, batch_offset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "inheritance states (64, 8)\n"
     ]
    }
   ],
   "source": [
    "class InheritanceStates:\n",
    "\n",
    "\t# inheritance states\n",
    "\t#\n",
    "\t# for parents:\n",
    "\t# (0, 1) -> deletion on parental1\n",
    "\t# (1, 0) -> deletion on parental2\n",
    "\t# (1, 1) -> normal\n",
    "\t# \n",
    "\t# for children:\n",
    "\t# (0, 0) -> m1p1\n",
    "\t# (0, 1) -> m1p2\n",
    "\t# (1, 0) -> m2p1\n",
    "\t# (1, 1) -> m2p2\n",
    "\n",
    "\tdef __init__(self, m):\n",
    "\t\tself.m = m\n",
    "\t\tif m >= 5:\n",
    "\t\t\tself.states = np.array(list(product(*([[0, 1]]*(2*m)))), dtype=np.int8)\n",
    "\t\telse:\n",
    "\t\t\tself.states = np.array([x for x in product(*([[0, 1]]*(2*m))) if x[4]==0 and x[5]==0], dtype=np.int8)\n",
    "\t\tprint('inheritance states', self.states.shape)\n",
    "\n",
    "\t\tself.p, self.state_len = self.states.shape\n",
    "\t\tself.state_to_index = dict([(tuple(x), i) for i, x in enumerate(self.states)])\n",
    "\n",
    "\tdef __getitem__(self, key):\n",
    "\t\treturn self.states[key]\n",
    "    \n",
    "\tdef index(self, s):\n",
    "\t\treturn self.state_to_index[s]\n",
    "\n",
    "\tdef __contains__(self, key):\n",
    "\t\treturn key in self.state_to_index\n",
    "    \n",
    "inheritance_states = InheritanceStates(m)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "transitions (64, 18)\n"
     ]
    }
   ],
   "source": [
    "class TransitionMatrix:\n",
    "\n",
    "\tdef __init__(self, inheritance_states, shift_costs):\n",
    "\n",
    "\t\t# transition matrix\n",
    "\t\tself.shift_costs = shift_costs\n",
    "\n",
    "\t\ttransitions = [[] for i in range(inheritance_states.p)]\n",
    "\t\ttransition_costs = [[] for i in range(inheritance_states.p)]\n",
    "\t\tfor i, state in enumerate(inheritance_states):\n",
    "\t\t\tfor delstate in list(product(*[[0, 1]]*4)):\n",
    "\t\t\t\tnew_state = tuple(delstate) + tuple(state[4:])\n",
    "\t\t\t\tif new_state in inheritance_states:\n",
    "\t\t\t\t\tnew_index = inheritance_states.index(new_state)\n",
    "\t\t\t\t\ttransitions[i].append(new_index)\n",
    "\t\t\t\t\ttransition_costs[i].append(sum([shift_costs[j] for j, (old_s, new_s) in enumerate(zip(state[:4], delstate)) if old_s != new_s]))\n",
    "\n",
    "\t\t\t# allow a single recombination event\n",
    "\t\t\tfor j in range(4, inheritance_states.state_len):\n",
    "\t\t\t\tnew_state = tuple(1-x if k == j else x for k, x in enumerate(state))\n",
    "\t\t\t\tif new_state in inheritance_states:\n",
    "\t\t\t\t\tnew_index = inheritance_states.index(new_state)\n",
    "\t\t\t\t\ttransitions[i].append(new_index)\n",
    "\t\t\t\t\ttransition_costs[i].append(shift_costs[j])\n",
    "\t\t            \n",
    "\t\tself.transitions = np.array(transitions)\n",
    "\t\tself.costs = np.array(transition_costs)\n",
    "\n",
    "\t\t# starting costs\n",
    "\t\tstarting_state = (1, 1, 1, 1, 0, 0)\n",
    "\t\tself.first_costs = np.zeros((inheritance_states.p,), dtype=int)\n",
    "\t\tfor i, ss in enumerate(starting_state):\n",
    "\t\t\tself.first_costs[inheritance_states[:, i] != ss] += shift_costs[i]\n",
    "\t\tprint('transitions', self.transitions.shape)\n",
    "transition_matrix = TransitionMatrix(inheritance_states, shift_costs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "genotypes (256, 4)\n"
     ]
    }
   ],
   "source": [
    "class Genotypes:\n",
    "\tdef __init__(self, m):\n",
    "\t\tself.genotypes = np.array(list(product(*[[-1, 0, 1, 2]]*m)), dtype=np.int8)\n",
    "\t\tself.genotype_to_index = dict([(tuple(x), i) for i, x in enumerate(self.genotypes)])\n",
    "\t\tself.q = self.genotypes.shape[0]\n",
    "\t\tprint('genotypes', self.genotypes.shape)\n",
    "\n",
    "\tdef __getitem__(self, key):\n",
    "\t\treturn self.genotypes[key]\n",
    "\n",
    "\tdef index(self, s):\n",
    "\t\treturn self.genotype_to_index[s]\n",
    "\n",
    "\tdef __contains__(self, key):\n",
    "\t\treturn key in self.genotype_to_index\n",
    "    \n",
    "genotypes = Genotypes(m)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "perfect matches (110, 4) Counter({4: 24, 2: 16, 8: 16, 1: 4, 15: 2, 16: 2})\n",
      "[4 4 4 4 3 2 3 2 1 2 1 2 1 1 1 1 3 3 2 2 2 1 1 1 0 1 0 0 0 0 0 0 1 1 2 2 0\n",
      " 0 1 0 0 0 0 1 0 0 0 0 1 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0]\n"
     ]
    }
   ],
   "source": [
    "class Loss:\n",
    "\tdef __init__(self, m, inheritance_states, genotypes, g_cost):\n",
    "\t\tself.g_cost = g_cost\n",
    "\n",
    "\t\tself.__build_perfect_matches__(m, inheritance_states)\n",
    "\t\tself.__build_loss_equivalence__(inheritance_states)\n",
    "\n",
    "\t\tself.genotypes = genotypes\n",
    "\t\tself.losses = np.zeros((int(p/4) if m>=5 else inheritance_states.p, genotypes.q), dtype=np.int8)\n",
    "\t\tself.already_calculated = np.zeros((genotypes.q,), dtype=bool)\n",
    "\t\t\n",
    "\tdef __call__(self, gen): \n",
    "\t\tgen_index = genotypes.index(gen)\n",
    "\t\tif not self.already_calculated[gen_index]:\n",
    "\t\t\ts = np.zeros((self.num_pm_gen,), dtype=np.int8)\n",
    "\t\t\tfor i, pm in enumerate(self.pm_gen):\n",
    "\t\t\t\ts[i] = sum([self.g_cost[(pred, obs)] for pred, obs in zip(pm, gen)])\n",
    "\t\t\t    \n",
    "\t\t\tfor i, indices in enumerate(self.pm_gen_indices):\n",
    "\t\t\t\tself.losses[i, gen_index] = np.min(s[indices])\n",
    "\t\t\tself.already_calculated[gen_index] = True\n",
    "\t\treturn self.losses[self.full_loss_indices, gen_index].astype(int)\n",
    "\n",
    "\tdef __build_perfect_matches__(self, m, inheritance_states):\n",
    "\t\t# perfect match genotypes\n",
    "\t\tpm_gen_to_index = dict()\n",
    "\t\tself.pm_gen_indices = []\n",
    "\t\tfor s in inheritance_states:\n",
    "\t\t\tif s[4] == 0 and s[5] == 0:\n",
    "\t\t\t\tanc_pos = [[-1] if s[i] == 0 else [0, 1] for i in range(4)]\n",
    "\t\t\t\tanc_variants = np.array(list(product(*anc_pos)), dtype=np.int8)\n",
    "\t\t\t\tpred_gens = np.zeros((anc_variants.shape[0], m), dtype=np.int8)\n",
    "\n",
    "\t\t\t\t# mom\n",
    "\t\t\t\t# deletion\n",
    "\t\t\t\tpred_gens[(anc_variants[:, 0]==-1) & (anc_variants[:, 1]==-1), 0] = -1\n",
    "\t\t\t\tpred_gens[(anc_variants[:, 0]==-1) & (anc_variants[:, 1]==0), 0] = 0\n",
    "\t\t\t\tpred_gens[(anc_variants[:, 0]==-1) & (anc_variants[:, 1]==1), 0] = 2\n",
    "\t\t\t\tpred_gens[(anc_variants[:, 0]==0) & (anc_variants[:, 1]==-1), 0] = 0\n",
    "\t\t\t\tpred_gens[(anc_variants[:, 0]==1) & (anc_variants[:, 1]==-1), 0] = 2\n",
    "\t\t\t\t# normal\n",
    "\t\t\t\tpred_gens[(anc_variants[:, 0]==0) & (anc_variants[:, 1]==0), 0] = 0\n",
    "\t\t\t\tpred_gens[(anc_variants[:, 0]==1) & (anc_variants[:, 1]==1), 0] = 2\n",
    "\t\t\t\tpred_gens[(anc_variants[:, 0]==0) & (anc_variants[:, 1]==1), 0] = 1\n",
    "\t\t\t\tpred_gens[(anc_variants[:, 0]==1) & (anc_variants[:, 1]==0), 0] = 1\n",
    "\n",
    "\t\t\t\t# dad\n",
    "\t\t\t\t# deletion\n",
    "\t\t\t\tpred_gens[(anc_variants[:, 2]==-1) & (anc_variants[:, 3]==-1), 1] = -1\n",
    "\t\t\t\tpred_gens[(anc_variants[:, 2]==-1) & (anc_variants[:, 3]==0), 1] = 0\n",
    "\t\t\t\tpred_gens[(anc_variants[:, 2]==-1) & (anc_variants[:, 3]==1), 1] = 2\n",
    "\t\t\t\tpred_gens[(anc_variants[:, 2]==0) & (anc_variants[:, 3]==-1), 1] = 0\n",
    "\t\t\t\tpred_gens[(anc_variants[:, 2]==1) & (anc_variants[:, 3]==-1), 1] = 2\n",
    "\t\t\t\t# normal\n",
    "\t\t\t\tpred_gens[(anc_variants[:, 2]==0) & (anc_variants[:, 3]==0), 1] = 0\n",
    "\t\t\t\tpred_gens[(anc_variants[:, 2]==1) & (anc_variants[:, 3]==1), 1] = 2\n",
    "\t\t\t\tpred_gens[(anc_variants[:, 2]==0) & (anc_variants[:, 3]==1), 1] = 1\n",
    "\t\t\t\tpred_gens[(anc_variants[:, 2]==1) & (anc_variants[:, 3]==0), 1] = 1\n",
    "\n",
    "\t\t\t\t# children\n",
    "\t\t\t\tfor index in range(m-2):\n",
    "\t\t\t\t\tmat, pat = s[(4+(2*index)):(6+(2*index))]\n",
    "\n",
    "\t\t\t\t\t# deletion\n",
    "\t\t\t\t\tpred_gens[(anc_variants[:, mat]==-1) & (anc_variants[:, 2+pat]==-1), 2+index] = -1\n",
    "\t\t\t\t\tpred_gens[(anc_variants[:, mat]==-1) & (anc_variants[:, 2+pat]==0), 2+index] = 0\n",
    "\t\t\t\t\tpred_gens[(anc_variants[:, mat]==-1) & (anc_variants[:, 2+pat]==1), 2+index] = 2\n",
    "\t\t\t\t\tpred_gens[(anc_variants[:, mat]==0) & (anc_variants[:, 2+pat]==-1), 2+index] = 0\n",
    "\t\t\t\t\tpred_gens[(anc_variants[:, mat]==1) & (anc_variants[:, 2+pat]==-1), 2+index] = 2\n",
    "\t\t\t\t\t# normal\n",
    "\t\t\t\t\tpred_gens[(anc_variants[:, mat]==0) & (anc_variants[:, 2+pat]==0), 2+index] = 0\n",
    "\t\t\t\t\tpred_gens[(anc_variants[:, mat]==1) & (anc_variants[:, 2+pat]==1), 2+index] = 2\n",
    "\t\t\t\t\tpred_gens[(anc_variants[:, mat]==0) & (anc_variants[:, 2+pat]==1), 2+index] = 1\n",
    "\t\t\t\t\tpred_gens[(anc_variants[:, mat]==1) & (anc_variants[:, 2+pat]==0), 2+index] = 1\n",
    "\n",
    "\t\t\t\tunique_pred_gens = set(map(tuple, pred_gens))\n",
    "\t\t\t\tfor pg in unique_pred_gens:\n",
    "\t\t\t\t\tif pg not in pm_gen_to_index:\n",
    "\t\t\t\t\t\tpm_gen_to_index[pg] = len(pm_gen_to_index)\n",
    "\t\t\t\tself.pm_gen_indices.append([pm_gen_to_index[pg] for pg in unique_pred_gens])\n",
    "\n",
    "\t\tself.num_pm_gen = len(pm_gen_to_index)\n",
    "\t\tself.pm_gen = np.zeros((self.num_pm_gen, m), dtype=np.int8)\n",
    "\t\tfor pm, i in pm_gen_to_index.items():\n",
    "\t\t\tself.pm_gen[i, :] = pm\n",
    "\n",
    "\t\tprint('perfect matches', self.pm_gen.shape, Counter([len(v) for v in self.pm_gen_indices]))\n",
    "\n",
    "\tdef __build_loss_equivalence__(self, inheritance_states):\n",
    "\t\t# losses are symmetrical to parental chromosome swaps\n",
    "\t\tself.full_loss_indices = np.zeros((inheritance_states.p,), dtype=int)\n",
    "\t\tloss_state_to_index = dict([(tuple(x), i) for i, x in enumerate(inheritance_states[(inheritance_states[:, 4]==0) & (inheritance_states[:, 5]==0), :])])\n",
    "\t\tfor i, s in enumerate(inheritance_states):\n",
    "\t\t\tnew_s = -np.ones((inheritance_states.state_len,), dtype=np.int8)\n",
    "\t\t\tif s[4] == 0:\n",
    "\t\t\t\tnew_s[:2] = s[:2]\n",
    "\t\t\t\tnew_s[np.arange(4, inheritance_states.state_len, 2)] = s[np.arange(4, inheritance_states.state_len, 2)]\n",
    "\t\t\telse:\n",
    "\t\t\t\tnew_s[:2] = s[[1, 0]]\n",
    "\t\t\t\tnew_s[np.arange(4, inheritance_states.state_len, 2)] = 1-s[np.arange(4, inheritance_states.state_len, 2)]\n",
    "\t\t\tif s[5] == 0:\n",
    "\t\t\t\tnew_s[2:4] = s[2:4]\n",
    "\t\t\t\tnew_s[np.arange(5, inheritance_states.state_len, 2)] = s[np.arange(5, inheritance_states.state_len, 2)]\n",
    "\t\t\telse:\n",
    "\t\t\t\tnew_s[2:4] = s[[3, 2]]\n",
    "\t\t\t\tnew_s[np.arange(5, inheritance_states.state_len, 2)] = 1-s[np.arange(5, inheritance_states.state_len, 2)]\n",
    "\t\t\tself.full_loss_indices[i] = loss_state_to_index[tuple(new_s)]\n",
    "            \n",
    "loss = Loss(m, inheritance_states, genotypes, g_cost)\n",
    "print(loss((0, 0, 0, 0)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chrom shape only SNPs (1004649,)\n"
     ]
    }
   ],
   "source": [
    "class WGSData:\n",
    "\tdef __init__(self, gen_files, coord_file, sample_file):\n",
    "\n",
    "\t\tself.gen_files = gen_files\n",
    "\n",
    "\t\twith open(sample_file, 'r') as f:\n",
    "\t\t\tsample_ids = [line.strip() for line in f]\n",
    "\t\tself.sample_id_to_index = dict([(sample_id, i) for i, sample_id in enumerate(sample_ids)])\n",
    "\n",
    "\t\t# use only \"cleaned\" variants - must be SNPs\n",
    "\t\tcoordinates = np.load(coord_file)\n",
    "\t\tself.snp_positions = coordinates[:, 1]\n",
    "\t\tself.snp_indices = coordinates[:, 2]==1\n",
    "\n",
    "\t\tself.snp_positions = self.snp_positions[self.snp_indices]\n",
    "\t\tprint('chrom shape only SNPs', self.snp_positions.shape)\n",
    "\n",
    "\tdef pull_data_for_individuals(self, individuals):\n",
    "\t\t#load from npz\n",
    "\t\tm = len(individuals)\n",
    "\t\tind_indices = [self.sample_id_to_index[x] for x in individuals]\n",
    "\t\t\n",
    "\t\tdata = sparse.hstack([sparse.load_npz('%s/%s' % (data_dir, gen_file))[ind_indices, :] for gen_file in self.gen_files]).A\n",
    "\t\tdata = data[:, self.snp_indices]\n",
    "\n",
    "\t\tn = 2*self.snp_positions.shape[0]+1\n",
    "\t\tfamily_genotypes = np.zeros((m, n), dtype=np.int8)\n",
    "\t\tfamily_genotypes[:, np.arange(1, n-1, 2)] = data\n",
    "\t\tfamily_genotypes[:, -2] = family_genotypes[:, -1]\n",
    "\n",
    "\t\t# if any family member is missing, set whole family to 0 - this has the effect of ignoring missing positions\n",
    "\t\t#family_genotypes[:, np.any(family_genotypes<0, axis=0)] = 0\n",
    "\t\t\n",
    "\t\t# if we see two missing entries in a row, mark the middle interval as possibly missing/possibly homref (-1)\n",
    "\t\tfamily_genotypes[family_genotypes<0] = -1\n",
    "\t\tfor i in range(m):\n",
    "\t\t\tdouble_missing = np.where((data[i, 1:]==-1) & (data[i, :-1]==-1))[0]\n",
    "\t\t\tfamily_genotypes[i, (2*double_missing)+2] = -1\n",
    "\n",
    "\t\tfamily_snp_positions = np.zeros((n, 2), dtype=np.int)\n",
    "\t\tfamily_snp_positions[0, 0] = 0\n",
    "\t\tfamily_snp_positions[np.arange(0, n-2, 2), 1] = self.snp_positions-1\n",
    "\t\tfamily_snp_positions[np.arange(1, n-1, 2), 0] = self.snp_positions-1\n",
    "\t\tfamily_snp_positions[np.arange(1, n-1, 2), 1] = self.snp_positions\n",
    "\t\tfamily_snp_positions[np.arange(2, n, 2), 0] = self.snp_positions\n",
    "\t\tfamily_snp_positions[-1, 1] = chrom_lengths[chrom]\n",
    "\n",
    "\t\t# remove unnecessary ref positions\n",
    "\t\thaslength = np.where(family_snp_positions[:, 0]!=family_snp_positions[:, 1])[0]\n",
    "\t\tfamily_genotypes = family_genotypes[:, haslength]\n",
    "\t\tfamily_snp_positions = family_snp_positions[haslength, :]\n",
    "\n",
    "\t\t# aggregate identical genotypes\n",
    "\t\trep_indices = np.where(np.any(family_genotypes[:, 1:]!=family_genotypes[:, :-1], axis=0))[0]\n",
    "\t\tn = rep_indices.shape[0]+1\n",
    "\n",
    "\t\tnew_family_genotypes = np.zeros((m, n), dtype=np.int8)\n",
    "\t\tnew_family_genotypes[:, :-1] = family_genotypes[:, rep_indices]\n",
    "\t\tnew_family_genotypes[:, -1] = family_genotypes[:, -1]\n",
    "\n",
    "\t\tnew_family_snp_positions = np.zeros((n, 2), dtype=np.int)\n",
    "\t\tnew_family_snp_positions[0, 0] = family_snp_positions[0, 0]\n",
    "\t\tnew_family_snp_positions[:-1, 1] = family_snp_positions[rep_indices, 1]\n",
    "\t\tnew_family_snp_positions[1:, 0] = family_snp_positions[rep_indices+1, 0]\n",
    "\t\tnew_family_snp_positions[-1, 1] = family_snp_positions[-1, 1]\n",
    "\n",
    "\t\tfamily_genotypes, family_snp_positions = new_family_genotypes, new_family_snp_positions\n",
    "\n",
    "\t\tmult_factor = family_snp_positions[:, 1] - family_snp_positions[:, 0]\n",
    "\n",
    "\t\treturn family_genotypes, family_snp_positions, mult_factor\n",
    "\n",
    "wgs_data = WGSData(gen_files, coord_file, sample_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "fkey, individuals = random.choice(families_of_this_size)\n",
    "\n",
    "# pull genotype data for this family\n",
    "family_genotypes, family_snp_positions, mult_factor = wgs_data.pull_data_for_individuals(individuals)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def viterbi_forward_sweep(family_genotypes, family_snp_positions, mult_factor, inheritance_states, transition_matrix, loss):\n",
    "\n",
    "\tm, n = family_genotypes.shape\n",
    "\tp, state_len = inheritance_states.p, inheritance_states.state_len\n",
    "\tv_cost = np.zeros((p, n), dtype=int)\n",
    "\t\t\n",
    "\t# forward sweep\n",
    "\tprev_time = time.time()\n",
    "\n",
    "\t# first step, break symmetry\n",
    "\t# we enforce that the chromosome starts with child1 (0, 0) and no deletions or duplications\n",
    "\tpos_gen = tuple(family_genotypes[:, 0])\n",
    "\tv_cost[:, 0] = mult_factor[0]*loss(pos_gen) + transition_matrix.first_costs\n",
    "\n",
    "\t# next steps\n",
    "\tfor j in range(1, n): \n",
    "\t\tpos_gen = tuple(family_genotypes[:, j])\n",
    "\t\tv_cost[:, j] = np.min(v_cost[transition_matrix.transitions, j-1] + transition_matrix.costs, axis=1) + mult_factor[j]*loss(pos_gen)\n",
    "\n",
    "\tprint('Forward sweep complete', time.time()-prev_time, 'sec') \n",
    "\n",
    "\treturn v_cost"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def viterbi_backward_sweep(v_cost, inheritance_states, transition_matrix):\n",
    "\n",
    "\t# backward sweep\n",
    "\tprev_time = time.time()\n",
    "\tn = v_cost.shape[1]\n",
    "\tfinal_states = -np.ones((inheritance_states.state_len, n), dtype=int)\n",
    "\t\n",
    "\t# choose best paths\n",
    "\t# we enforce that the chromosome ends with no deletions\n",
    "\tnum_forks = 0\n",
    "\tno_delstates = np.all(inheritance_states[:, :4]==1, axis=1)\n",
    "\tmin_value = np.min(v_cost[no_delstates, -1])\n",
    "\tpaths = np.where((v_cost[:, -1]==min_value) & no_delstates)[0]\n",
    "\tprint('Num solutions', paths.shape, inheritance_states[paths, :])\n",
    "\n",
    "\t# combine path states into a single state (unknown values represented with -1)\n",
    "\tif paths.shape[0] == 1:\n",
    "\t\tfinal_states[:, -1] = inheritance_states[paths[0], :]\n",
    "\telse:\n",
    "\t\tnum_forks += 1\n",
    "\t\tpath_states = inheritance_states[paths, :]\n",
    "\t\tknown_indices = np.all(path_states == path_states[0, :], axis=0)\n",
    "\t\tfinal_states[known_indices, -1] = path_states[0, known_indices]\n",
    "\n",
    "\t# now work backwards\n",
    "\tfor j in reversed(range(n-1)):\n",
    "\t\t# traceback\n",
    "\t\ttotal_cost = v_cost[transition_matrix.transitions[paths, :], j] + transition_matrix.costs[paths, :]\n",
    "\t\tmin_value = np.min(total_cost, axis=1)\n",
    "\t\tnew_paths = set()\n",
    "\t\tfor i, k in enumerate(paths):\n",
    "\t\t\tmin_indices = transition_matrix.transitions[k, np.where(total_cost[i, :] == min_value[i])[0]]\t\n",
    "\t\t\tnew_paths.update(min_indices.tolist())\n",
    "\t\tpaths = np.asarray(list(new_paths), dtype=int)\n",
    "\n",
    "\t\t# combine path states a single state (unknown values represented with -1)\n",
    "\t\tif paths.shape[0] == 1:\n",
    "\t\t\tfinal_states[:, j] = inheritance_states[paths[0], :]\n",
    "\t\telse:\n",
    "\t\t\tnum_forks += 1\n",
    "\t\t\tpath_states = inheritance_states[paths, :]\n",
    "\t\t\tknown_indices = np.all(path_states == path_states[0, :], axis=0)\n",
    "\t\t\tfinal_states[known_indices, j] = path_states[0, known_indices]\n",
    "\n",
    "\tprint('Num positions in fork', num_forks)\n",
    "\tprint('Backward sweep complete', time.time()-prev_time, 'sec') \n",
    "\t\n",
    "\treturn final_states"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def mask_states(family_genotypes, final_states, inheritance_states, loss, smooth, error_rate):\n",
    "\n",
    "\t# Now, do masking\n",
    "\tprev_time = time.time()\n",
    "\n",
    "\tm, n = family_genotypes.shape\n",
    "\tprint(family_genotypes.shape, final_states.shape)\n",
    "\tfit = -np.ones((n,), dtype=int)\n",
    "\tprev_state = None\n",
    "\tprev_state_indices = None\n",
    "\tfor j in range(n): \n",
    "\t\tpos_gen = tuple(family_genotypes[:, j])\n",
    "\t\tcurrent_state = tuple(final_states[:, j])\n",
    "\n",
    "\t\tif current_state != prev_state:\n",
    "\t\t\tprev_state = current_state\n",
    "\t\t\tnum_unknowns = len([x for x in current_state if x == -1])\n",
    "\t\t\tif num_unknowns>0:\n",
    "\t\t\t\tprev_state_indices = []\n",
    "\t\t\t\tfor poss_itr in [iter(x) for x in product(*([[0, 1]]*num_unknowns))]:\n",
    "\t\t\t\t\tposs_state = tuple([x if x != -1 else next(poss_itr) for x in current_state])\n",
    "\t\t\t\t\tprev_state_indices.append(inheritance_states.index(poss_state))\n",
    "\t\t\telse:\n",
    "\t\t\t\tprev_state_indices = [inheritance_states.index(tuple(final_states[:, j]))]\n",
    "\n",
    "\t\tfit[j] = mult_factor[j]*np.min(loss(pos_gen)[prev_state_indices])\n",
    "\n",
    "\n",
    "\tc = np.convolve(fit/m, np.ones(smooth,), mode='same')\n",
    "\tmasked = (c>(error_rate*smooth)).astype(np.int8)\n",
    "\tprint('Percent masked', 100*np.sum(masked)/n)\n",
    "\n",
    "\t# if a parental chromosome isn't inherited, then we don't know if it has a deletion\n",
    "\tmaternal_indices = range(4, inheritance_states.state_len, 2)\n",
    "\tpaternal_indices = range(5, inheritance_states.state_len, 2)\n",
    "\n",
    "\tm1_ninh = np.all(final_states[maternal_indices, :]!=0, axis=0)\n",
    "\tm2_ninh = np.all(final_states[maternal_indices, :]!=1, axis=0)\n",
    "\tp1_ninh = np.all(final_states[paternal_indices, :]!=0, axis=0)\n",
    "\tp2_ninh = np.all(final_states[paternal_indices, :]!=1, axis=0)\n",
    "\n",
    "\tfinal_states[0, m1_ninh] = -1\n",
    "\tfinal_states[1, m2_ninh] = -1\n",
    "\tfinal_states[2, p1_ninh] = -1\n",
    "\tfinal_states[3, p2_ninh] = -1\n",
    "\tfinal_states = np.append(final_states, masked[np.newaxis, :], axis=0)\n",
    "\tprint('Percent never inherited', 100*np.sum(m1_ninh)/n, 100*np.sum(m2_ninh)/n, 100*np.sum(p1_ninh)/n, 100*np.sum(p2_ninh)/n)\n",
    "\n",
    "\tprint('Masking complete', time.time()-prev_time, 'sec') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Forward sweep complete 6.700500965118408 sec\n",
      "Num solutions (1,) [[1 1 1 1 0 0 1 1]]\n",
      "Num positions in fork 202\n",
      "Backward sweep complete 7.268156051635742 sec\n",
      "(4, 158885) (8, 158885)\n",
      "Percent masked 20.809390439626146\n",
      "Percent never inherited 0.0 60.06986184976555 0.0 85.186140919533\n",
      "Masking complete 4.246056079864502 sec\n"
     ]
    }
   ],
   "source": [
    "# forward sweep\n",
    "v_cost = viterbi_forward_sweep(family_genotypes, family_snp_positions, mult_factor, inheritance_states, transition_matrix, loss)\n",
    "\n",
    "# backward sweep\n",
    "final_states = viterbi_backward_sweep(v_cost, inheritance_states, transition_matrix)\n",
    "\n",
    "# mask messy areas\n",
    "mask_states(family_genotypes, final_states, inheritance_states, loss, smooth, error_rate)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA2QAAAEyCAYAAACVoBMLAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4yLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvhp/UCwAAEu5JREFUeJzt3WusZWdZB/D/4zktl4IUyoiVGW0h\nhKQxgeIJwWCMFoWChPqBxBIv3MwEjQa8xBRJTPzoJd4ioZkAiopQrKCEqLUqhphI8bSUUnqRaQGZ\nBuxQ5FYTasfHD3sNHMaZnn3K3uc9s8/vl6yctd71nrWfOW/eOfs/a+13qrsDAADA7vuW0QUAAADs\nVwIZAADAIAIZAADAIAIZAADAIAIZAADAIAIZAADAIAIZAADAIAIZAADAIAIZAADAIOvLuOgTn7DW\nFx06ZxmXBgAA2PNuvOWrn+vuA9v1W0ogu+jQOfnQdYeWcWkAAIA9b+3Co5+ap59HFgEAAAYRyAAA\nAAYRyAAAAAYRyAAAAAYRyAAAAAYRyAAAAAYRyAAAAAYRyAAAAAYRyAAAAAYRyAAAAAYRyAAAAAYR\nyAAAAAaZK5BV1eVVdWdVHa2qq5ZdFAAAwH6wbSCrqrUkb0zywiSXJHlZVV2y7MIAAABW3Tx3yJ6d\n5Gh3393dDyR5Z5IrllsWAADA6psnkD05yae3HB+b2r5BVR2uqs2q2jx+34lF1QcAALCyFraoR3cf\n6e6N7t44cMHaoi4LAACwsuYJZPckObTl+ODUBgAAwDdhnkD2b0meVlUXV9W5Sa5M8t7llgUAALD6\n1rfr0N0PVtXPJbkuyVqSt3b3x5ZeGQAAwIrbNpAlSXf/TZK/WXItAAAA+8rCFvUAAABgZwQyAACA\nQQQyAACAQQQyAACAQQQyAACAQQQyAACAQQQyAACAQQQyAACAQQQyAACAQQQyAACAQQQyAACAQQQy\nAACAQQQyAACAQQQyAACAQQQyAACAQQQyAACAQQQyAACAQQQyAACAQQQyAACAQQQyAACAQQQyAACA\nQQQyAACAQQQyAACAQQQyAACAQQQyAACAQbYNZFX11qq6t6pu3Y2CAAAA9ot57pD9cZLLl1wHAADA\nvrNtIOvuDyT5/C7UAgAAsK8s7DNkVXW4qjaravP4fScWdVkAAICVtbBA1t1HunujuzcOXLC2qMsC\nAACsLKssAgAADCKQAQAADDLPsvfvSPKvSZ5eVceq6tXLLwsAAGD1rW/XobtfthuFAAAA7DceWQQA\nABhEIAMAABhEIAMAABhEIAMAABhEIAMAABhEIAMAABhEIAMAABhEIAMAABhEIAMAABhEIAMAABhE\nIAMAABhEIAMAABhEIAMAABhEIAMAABhEIAMAABhEIAMAABhEIAMAABhEIAMAABhEIAMAABhEIAMA\nABhEIAMAABhEIAMAABhEIAMAABhEIAMAABhEIAMAABhk20BWVYeq6v1VdVtVfayqXrsbhQEAAKy6\n9Tn6PJjkl7r7pqp6bJIbq+r67r5tybUBAACstG3vkHX3Z7r7pmn/y0luT/LkZRcGAACw6nb0GbKq\nuijJpUluOM25w1W1WVWbx+87sZjqAAAAVtjcgayqHpPkL5O8rru/dOr57j7S3RvdvXHggrVF1ggA\nALCS5gpkVXVOZmHs7d397uWWBAAAsD/Ms8piJXlLktu7+3eWXxIAAMD+MM8dsucm+ckkl1XVzdP2\noiXXBQAAsPK2Xfa+u/8lSe1CLQAAAPvKjlZZBAAAYHEEMgAAgEEEMgAAgEEEMgAAgEEEMgAAgEEE\nMgAAgEEEMgAAgEEEMgAAgEEEMgAAgEEEMgAAgEEEMgAAgEEEMgAAgEEEMgAAgEEEMgAAgEEEMgAA\ngEEEMgAAgEEEMgAAgEEEMgAAgEEEMgAAgEEEMgAAgEEEMgAAgEEEMgAAgEEEMgAAgEEEMgAAgEEE\nMgAAgEG2DWRV9ciq+lBVfaSqPlZVv74bhQEAAKy69Tn6fDXJZd39lao6J8m/VNXfdvcHl1wbAADA\nSts2kHV3J/nKdHjOtPUyiwIAANgP5voMWVWtVdXNSe5Ncn1333CaPoerarOqNo/fd2LRdQIAAKyc\nuQJZd5/o7mcmOZjk2VX13afpc6S7N7p748AFa4uuEwAAYOXsaJXF7v5CkvcnuXw55QAAAOwf86yy\neKCqzp/2H5Xkh5PcsezCAAAAVt08qyxemORtVbWWWYB7V3e/b7llAQAArL55Vlm8Jcmlu1ALAADA\nvrKjz5ABAACwOAIZAADAIAIZAADAIAIZAADAIAIZAADAIAIZAADAIAIZAADAIAIZAADAIAIZAADA\nIAIZAADAIAIZAADAIOvLuOhH/+tAnnrNa5ZxaQAAgLPAL8/Vyx0yAACAQQQyAACAQQQyAACAQQQy\nAACAQQQyAACAQQQyAACAQQQyAACAQQQyAACAQQQyAACAQQQyAACAQQQyAACAQQQyAACAQQQyAACA\nQeYOZFW1VlUfrqr3LbMgAACA/WInd8hem+T2ZRUCAACw38wVyKrqYJIfSfLm5ZYDAACwf8x7h+z3\nkvxKkv89U4eqOlxVm1W1eeIr9y+kOAAAgFW2bSCrqhcnube7b3yoft19pLs3untj7THnLaxAAACA\nVTXPHbLnJnlJVX0yyTuTXFZVf7bUqgAAAPaBbQNZd7++uw9290VJrkzyT939E0uvDAAAYMX5f8gA\nAAAGWd9J5+7+5yT/vJRKAAAA9hl3yAAAAAYRyAAAAAYRyAAAAAYRyAAAAAYRyAAAAAYRyAAAAAYR\nyAAAAAYRyAAAAAYRyAAAAAYRyAAAAAYRyAAAAAYRyAAAAAYRyAAAAAYRyAAAAAYRyAAAAAYRyAAA\nAAYRyAAAAAYRyAAAAAYRyAAAAAYRyAAAAAYRyAAAAAYRyAAAAAYRyAAAAAYRyAAAAAYRyAAAAAZZ\nn6dTVX0yyZeTnEjyYHdvLLMoAACA/WCuQDb5we7+3NIqAQAA2Gc8sggAADDIvIGsk/x9Vd1YVYdP\n16GqDlfVZlVtnvjK/YurEAAAYEXN+8ji93X3PVX1bUmur6o7uvsDWzt095EkR5LkEd95qBdcJwAA\nwMqZ6w5Zd98zfb03yXuSPHuZRQEAAOwH2wayqjqvqh57cj/J85PcuuzCAAAAVt08jyw+Kcl7qupk\n/z/v7r9balUAAAD7wLaBrLvvTvKMXagFAABgX7HsPQAAwCACGQAAwCACGQAAwCACGQAAwCACGQAA\nwCACGQAAwCACGQAAwCACGQAAwCACGQAAwCACGQAAwCACGQAAwCDV3Qu/6MYzHtkfuu7Qwq8LsEhP\nveY1o0sAABbkrh+7enQJ32DtwqM3dvfGdv3cIQMAABhEIAMAABhEIAMAABhEIAMAABhEIAMAABhE\nIAMAABhEIAMAABhEIAMAABhEIAMAABhEIAMAABhEIAMAABhEIAMAABhEIAMAABhkrkBWVedX1bVV\ndUdV3V5V37vswgAAAFbd+pz9fj/J33X3S6vq3CSPXmJNAAAA+8K2gayqHpfk+5O8Ikm6+4EkDyy3\nLAAAgNU3zyOLFyc5nuSPqurDVfXmqjrv1E5VdbiqNqtq8/h9JxZeKAAAwKqZJ5CtJ3lWkjd196VJ\n7k9y1amduvtId29098aBC9YWXCYAAMDqmSeQHUtyrLtvmI6vzSygAQAA8E3YNpB192eTfLqqnj41\nPS/JbUutCgAAYB+Yd5XFn0/y9mmFxbuTvHJ5JQEAAOwPcwWy7r45ycaSawEAANhX5vqPoQEAAFg8\ngQwAAGAQgQwAAGAQgQwAAGAQgQwAAGAQgQwAAGAQgQwAAGAQgQwAAGAQgQwAAGAQgQwAAGAQgQwA\nAGCQ9dEFsFhPveY1o0sAAIBd91Dvg+/6sat3sZKdcYcMAABgEIEMAABgEIEMAABgEIEMAABgEIEM\nAABgEIEMAABgEIEMAABgEIEMAABgEIEMAABgEIEMAABgEIEMAABgEIEMAABgEIEMAABgkG0DWVU9\nvapu3rJ9qapetxvFAQAArLL17Tp0951JnpkkVbWW5J4k71lyXQAAACtvp48sPi/JXd39qWUUAwAA\nsJ/sNJBdmeQdpztRVYerarOqNo/fd+KbrwwAAGDFzR3IqurcJC9J8henO9/dR7p7o7s3Dlywtqj6\nAAAAVtZO7pC9MMlN3f2fyyoGAABgP9lJIHtZzvC4IgAAADs3VyCrqvOS/HCSdy+3HAAAgP1j22Xv\nk6S7709ywZJrAQAA2Fd2usoiAAAACyKQAQAADCKQAQAADCKQAQAADCKQAQAADCKQAQAADCKQAQAA\nDCKQAQAADCKQAQAADCKQAQAADCKQAQAADFLdvfiLVn05yZ0LvzCL9MQknxtdBA/JGJ0djNPeZ4z2\nPmN0djBOe58x2lu+q7sPbNdpfUkvfmd3byzp2ixAVW0ao73NGJ0djNPeZ4z2PmN0djBOe58xOjt5\nZBEAAGAQgQwAAGCQZQWyI0u6LotjjPY+Y3R2ME57nzHa+4zR2cE47X3G6Cy0lEU9AAAA2J5HFgEA\nAAYRyAAAAAZZaCCrqsur6s6qOlpVVy3y2vx/VXWoqt5fVbdV1ceq6rVT+xOq6vqq+vj09fFTe1XV\nH0zjc0tVPWvLtV4+9f94Vb18S/v3VNVHp+/5g6qq3f+Tnv2qaq2qPlxV75uOL66qG6af6zVVde7U\n/ojp+Oh0/qIt13j91H5nVb1gS7t5twBVdX5VXVtVd1TV7VX1vebS3lJVvzD9XXdrVb2jqh5pLo1X\nVW+tqnur6tYtbUufO2d6Df6/M4zRb01/391SVe+pqvO3nNvRHHk485BvdLox2nLul6qqq+qJ07F5\ntGq6eyFbkrUkdyV5SpJzk3wkySWLur7ttD/zC5M8a9p/bJJ/T3JJkt9MctXUflWS35j2X5Tkb5NU\nkuckuWFqf0KSu6evj5/2Hz+d+9DUt6bvfeHoP/fZuCX5xSR/nuR90/G7klw57V+d5Gem/Z9NcvW0\nf2WSa6b9S6Y59YgkF09zbc28W+gYvS3JT0/75yY531zaO1uSJyf5RJJHTcfvSvIKc2n8luT7kzwr\nya1b2pY+d870Gra5x+j5Sdan/d/YMkY7niM7nYe2+cZoaj+U5Lokn0ryxKnNPFqxbZF3yJ6d5Gh3\n393dDyR5Z5IrFnh9TtHdn+num6b9Lye5PbM3LVdk9uYy09cfnfavSPInPfPBJOdX1YVJXpDk+u7+\nfHf/V5Lrk1w+nfvW7v5gz2bqn2y5FnOqqoNJfiTJm6fjSnJZkmunLqeO0cmxuzbJ86b+VyR5Z3d/\ntbs/keRoZnPOvFuAqnpcZr8M35Ik3f1Ad38h5tJes57kUVW1nuTRST4Tc2m47v5Aks+f0rwbc+dM\nr8EpTjdG3f333f3gdPjBJAen/R3NkYf5O41TnGEeJcnvJvmVJFtX4TOPVswiA9mTk3x6y/GxqY1d\nMD0GcGmSG5I8qbs/M536bJInTftnGqOHaj92mnZ25vcy+8v0f6fjC5J8Ycsvwq0/16+NxXT+i1P/\nnY4dO3NxkuNJ/qhmj5a+uarOi7m0Z3T3PUl+O8l/ZBbEvpjkxphLe9VuzJ0zvQY796rM7pokOx+j\nh/M7jTlU1RVJ7unuj5xyyjxaMRb1WAFV9Zgkf5nkdd39pa3npn8J8X8bDFJVL05yb3ffOLoWHtJ6\nZo+KvKm7L01yf2aPbnyNuTTW9LmGKzILz9+R5Lwklw8tirnsxtwxPx++qnpDkgeTvH10LXxdVT06\nya8m+bXdek3zaJxFBrJ7MnvO9aSDUxtLVFXnZBbG3t7d756a/3O6PZ3p671T+5nG6KHaD56mnfk9\nN8lLquqTmT3ecVmS38/s8YL1qc/Wn+vXxmI6/7gk92XnY8fOHEtyrLtvmI6vzSygmUt7xw8l+UR3\nH+/u/0ny7szml7m0N+3G3DnTazCnqnpFkhcn+fHpzXiy8zG6Lzufh2zvqZn9A9RHpvcQB5PcVFXf\nHvNo5SwykP1bkqdNK+2cm9mHN9+7wOtziuk57Lckub27f2fLqfcmObmyzsuT/PWW9p+aVud5TpIv\nTrepr0vy/Kp6/PSv0M9Pct107ktV9ZzptX5qy7WYQ3e/vrsPdvdFmc2Jf+ruH0/y/iQvnbqdOkYn\nx+6lU/+e2q+cVqy6OMnTMvuArnm3AN392SSfrqqnT03PS3JbzKW95D+SPKeqHj39DE+Okbm0N+3G\n3DnTazCHqro8s8fpX9Ld/73l1I7myDSvdjoP2UZ3f7S7v627L5reQxzLbCG3z8Y8Wj29oNVBpvn1\nosxW+rsryRsWeW3baX/e35fZreVbktw8bS/K7Pnsf0zy8ST/kOQJU/9K8sZpfD6aZGPLtV6V2Qd3\njyZ55Zb2jSS3Tt/zh0lq9J/7bN2S/EC+vsriUzL7BXc0yV8kecTU/sjp+Oh0/ilbvv8N0zjcmS0r\n9Jl3CxufZybZnObTX2W2QpW5tIe2JL+e5I7p5/inma0CZy6NH5d3ZPa5vv/J7E3jq3dj7pzpNWxz\nj9HRzD5vdPL9w9Vb+u9ojjyceWjbfoxOOf/JfH2VRfNoxbaTgwEAAMAus6gHAADAIAIZAADAIAIZ\nAADAIAIZAADAIAIZAADAIAIZAADAIAIZAADAIP8HSASmpN2rWt8AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 1080x360 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "plt.figure(figsize=(15, 5))\n",
    "plt.imshow(final_states, aspect='auto')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA2QAAAEyCAYAAACVoBMLAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4yLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvhp/UCwAAEylJREFUeJzt3X2sZWddL/Dvz3PaAgUplBELM9pC\nCEljYss9IRiM0aJQkFD/ILFEr+BLJnpzb+i9GFMkuYl/eu+Nb9HYTABfkRcrKCHqWBVDTC7F01JK\n6YtMS5EpYIcibzWhdvz5x14Dh3GmZ5+y93nO7PP5JCtnrWc9Z+3fnCfPnP2dtfYz1d0BAABg933L\n6AIAAAD2K4EMAABgEIEMAABgEIEMAABgEIEMAABgEIEMAABgEIEMAABgEIEMAABgEIEMAABgkPVl\nXPQZT1/rSw+dt4xLAwAA7Hm33P7Vz3X3ge36LSWQXXrovHzo6KFlXBoAAGDPW7vk2Cfn6eeRRQAA\ngEEEMgAAgEEEMgAAgEEEMgAAgEEEMgAAgEEEMgAAgEEEMgAAgEEEMgAAgEEEMgAAgEEEMgAAgEEE\nMgAAgEEEMgAAgEHmCmRVdXVV3VNVx6rq+mUXBQAAsB9sG8iqai3JbyV5eZLLk7ymqi5fdmEAAACr\nbp47ZC9Mcqy77+vuR5K8I8k1yy0LAABg9c0TyJ6d5FNbjo9Pbd+gqg5X1WZVbZ546OSi6gMAAFhZ\nC1vUo7uPdPdGd28cuHhtUZcFAABYWfMEsgeSHNpyfHBqAwAA4JswTyD7hyTPq6rLqur8JNcmee9y\nywIAAFh969t16O5Hq+q/JzmaZC3JW7v7Y0uvDAAAYMVtG8iSpLv/PMmfL7kWAACAfWVhi3oAAACw\nMwIZAADAIAIZAADAIAIZAADAIAIZAADAIAIZAADAIAIZAADAIAIZAADAIAIZAADAIAIZAADAIAIZ\nAADAIOujCwAY5WXPumJ0CQDAghz99G2jS3hc3CEDAAAYRCADAAAYRCADAAAYRCADAAAYRCADAAAY\nRCADAAAYRCADAAAYRCADAAAYRCADAAAYRCADAAAYRCADAAAYRCADAAAYRCADAAAYZNtAVlVvraoH\nq+qO3SgIAABgv5jnDtnvJrl6yXUAAADsO9sGsu7+QJLP70ItAAAA+8rCPkNWVYerarOqNk88dHJR\nlwUAAFhZCwtk3X2kuze6e+PAxWuLuiwAAMDKssoiAADAIAIZAADAIPMse//2JP8/yfOr6nhV/fTy\nywIAAFh969t16O7X7EYhAAAA+41HFgEAAAYRyAAAAAYRyAAAAAYRyAAAAAYRyAAAAAYRyAAAAAYR\nyAAAAAYRyAAAAAYRyAAAAAYRyAAAAAYRyAAAAAZZH10AwChHP33b6BIAgH3OHTIAAIBBBDIAAIBB\nBDIAAIBBBDIAAIBBBDIAAIBBBDIAAIBBBDIAAIBBBDIAAIBBBDIAAIBBBDIAAIBBBDIAAIBBBDIA\nAIBBBDIAAIBBtg1kVXWoqt5fVXdW1ceq6vW7URgAAMCqW5+jz6NJ3tDdt1bVU5LcUlU3dfedS64N\nAABgpW17h6y7P9Pdt077X05yV5JnL7swAACAVbejz5BV1aVJrkxy8xnOHa6qzaraPPHQycVUBwAA\nsMLmDmRV9eQkf5Lkuu7+0unnu/tId29098aBi9cWWSMAAMBKmiuQVdV5mYWxt3X3u5dbEgAAwP4w\nzyqLleQtSe7q7l9ZfkkAAAD7wzx3yF6c5L8muaqqbpu2Vyy5LgAAgJW37bL33f33SWoXagEAANhX\ndrTKIgAAAIsjkAEAAAwikAEAAAwikAEAAAwikAEAAAwikAEAAAwikAEAAAwikAEAAAwikAEAAAwi\nkAEAAAwikAEAAAyyProAFutlz7pidAkAALCnHP30baNLOCt3yAAAAAYRyAAAAAYRyAAAAAYRyAAA\nAAYRyAAAAAYRyAAAAAYRyAAAAAYRyAAAAAYRyAAAAAYRyAAAAAYRyAAAAAYRyAAAAAYRyAAAAAbZ\nNpBV1ROq6kNV9ZGq+lhV/dJuFAYAALDq1ufo89UkV3X3V6rqvCR/X1V/0d0fXHJtAAAAK23bQNbd\nneQr0+F509bLLAoAAGA/mOszZFW1VlW3JXkwyU3dffMZ+hyuqs2q2jzx0MlF1wkAALBy5gpk3X2y\nu69IcjDJC6vqu87Q50h3b3T3xoGL1xZdJwAAwMrZ0SqL3f2FJO9PcvVyygEAANg/5lll8UBVXTTt\nPzHJDyW5e9mFAQAArLp5Vlm8JMnvVdVaZgHuXd39vuWWBQAAsPrmWWXx9iRX7kItAAAA+8qOPkMG\nAADA4ghkAAAAgwhkAAAAgwhkAAAAgwhkAAAAgwhkAAAAgwhkAAAAgwhkAAAAgwhkAAAAgwhkAAAA\ngwhkAAAAg1R3L/yiF3zHoX7WG65b+HUBAADOBfdf9/O3dPfGdv3cIQMAABhEIAMAABhEIAMAABhE\nIAMAABhEIAMAABhEIAMAABhEIAMAABhEIAMAABhEIAMAABhEIAMAABhEIAMAABhEIAMAABhEIAMA\nABhk7kBWVWtV9eGqet8yCwIAANgvdnKH7PVJ7lpWIQAAAPvNXIGsqg4m+eEkb15uOQAAAPvHvHfI\nfi3JLyT597N1qKrDVbVZVZsnv/LwQooDAABYZdsGsqp6ZZIHu/uWx+rX3Ue6e6O7N9aefOHCCgQA\nAFhV89whe3GSV1XV/UnekeSqqvrDpVYFAACwD2wbyLr7jd19sLsvTXJtkr/t7h9femUAAAArzv9D\nBgAAMMj6Tjp3998l+bulVAIAALDPuEMGAAAwiEAGAAAwiEAGAAAwiEAGAAAwiEAGAAAwiEAGAAAw\niEAGAAAwiEAGAAAwiEAGAAAwiEAGAAAwiEAGAAAwiEAGAAAwiEAGAAAwiEAGAAAwiEAGAAAwiEAG\nAAAwiEAGAAAwiEAGAAAwiEAGAAAwiEAGAAAwiEAGAAAwiEAGAAAwiEAGAAAwiEAGAAAwiEAGAAAw\nyPo8narq/iRfTnIyyaPdvbHMogAAAPaDuQLZ5Ae6+3NLqwQAAGCf8cgiAADAIPMGsk7yV1V1S1Ud\nPlOHqjpcVZtVtXnyKw8vrkIAAIAVNe8ji9/b3Q9U1bcluamq7u7uD2zt0N1HkhxJkgu+41AvuE4A\nAICVM9cdsu5+YPr6YJL3JHnhMosCAADYD7YNZFV1YVU95dR+kpcmuWPZhQEAAKy6eR5ZfGaS91TV\nqf5/1N1/udSqAAAA9oFtA1l335fku3ehFgAAgH3FsvcAAACDCGQAAACDCGQAAACDCGQAAACDCGQA\nAACDCGQAAACDCGQAAACDCGQAAACDCGQAAACDCGQAAACDCGQAAACDVHcv/KIb3/2E/tDRQwu/LsAi\nPfedPzu6BABgQe790RtGl/AN1i45dkt3b2zXzx0yAACAQQQyAACAQQQyAACAQQQyAACAQQQyAACA\nQQQyAACAQQQyAACAQQQyAACAQQQyAACAQQQyAACAQQQyAACAQQQyAACAQQQyAACAQeYKZFV1UVXd\nWFV3V9VdVfU9yy4MAABg1a3P2e/Xk/xld7+6qs5P8qQl1gQAALAvbBvIquqpSb4vyeuSpLsfSfLI\ncssCAABYffM8snhZkhNJfqeqPlxVb66qC0/vVFWHq2qzqjZPPHRy4YUCAACsmnkC2XqSFyT57e6+\nMsnDSa4/vVN3H+nuje7eOHDx2oLLBAAAWD3zBLLjSY53983T8Y2ZBTQAAAC+CdsGsu7+bJJPVdXz\np6aXJLlzqVUBAADsA/Ousvg/krxtWmHxviQ/ubySAAAA9oe5All335ZkY8m1AAAA7Ctz/cfQAAAA\nLJ5ABgAAMIhABgAAMIhABgAAMIhABgAAMIhABgAAMIhABgAAMIhABgAAMIhABgAAMIhABgAAMIhA\nBgAAMMj66AJYrOe+82dHlwAAALvusd4H3/ujN+xiJTvjDhkAAMAgAhkAAMAgAhkAAMAgAhkAAMAg\nAhkAAMAgAhkAAMAgAhkAAMAgAhkAAMAgAhkAAMAgAhkAAMAgAhkAAMAgAhkAAMAgAhkAAMAg2way\nqnp+Vd22ZftSVV23G8UBAACssvXtOnT3PUmuSJKqWkvyQJL3LLkuAACAlbfTRxZfkuTe7v7kMooB\nAADYT3YayK5N8vYznaiqw1W1WVWbJx46+c1XBgAAsOLmDmRVdX6SVyX54zOd7+4j3b3R3RsHLl5b\nVH0AAAArayd3yF6e5Nbu/udlFQMAALCf7CSQvSZneVwRAACAnZsrkFXVhUl+KMm7l1sOAADA/rHt\nsvdJ0t0PJ7l4ybUAAADsKztdZREAAIAFEcgAAAAGEcgAAAAGEcgAAAAGEcgAAAAGEcgAAAAGEcgA\nAAAGEcgAAAAGEcgAAAAGEcgAAAAGEcgAAAAGqe5e/EWrvpzknoVfmEV6RpLPjS6Cx2SMzg3Gae8z\nRnufMTo3GKe9zxjtLd/Z3Qe267S+pBe/p7s3lnRtFqCqNo3R3maMzg3Gae8zRnufMTo3GKe9zxid\nmzyyCAAAMIhABgAAMMiyAtmRJV2XxTFGe58xOjcYp73PGO19xujcYJz2PmN0DlrKoh4AAABszyOL\nAAAAgwhkAAAAgyw0kFXV1VV1T1Udq6rrF3lt/rOqOlRV76+qO6vqY1X1+qn96VV1U1V9fPr6tKm9\nquo3pvG5vapesOVar536f7yqXrul/b9U1Uen7/mNqqrd/5Oe+6pqrao+XFXvm44vq6qbp5/rO6vq\n/Kn9gun42HT+0i3XeOPUfk9VvWxLu3m3AFV1UVXdWFV3V9VdVfU95tLeUlX/c/q77o6qentVPcFc\nGq+q3lpVD1bVHVvalj53zvYa/GdnGaP/O/19d3tVvaeqLtpybkdz5PHMQ77RmcZoy7k3VFVX1TOm\nY/No1XT3QrYka0nuTfKcJOcn+UiSyxd1fdsZf+aXJHnBtP+UJP+Y5PIk/yfJ9VP79Ul+edp/RZK/\nSFJJXpTk5qn96Unum74+bdp/2nTuQ1Pfmr735aP/3OfiluR/JfmjJO+bjt+V5Npp/4YkPzft/7ck\nN0z71yZ557R/+TSnLkhy2TTX1sy7hY7R7yX5mWn//CQXmUt7Z0vy7CSfSPLE6fhdSV5nLo3fknxf\nkhckuWNL29Lnztlewzb3GL00yfq0/8tbxmjHc2Sn89A23xhN7YeSHE3yySTPmNrMoxXbFnmH7IVJ\njnX3fd39SJJ3JLlmgdfnNN39me6+ddr/cpK7MnvTck1mby4zff2Raf+aJL/fMx9MclFVXZLkZUlu\n6u7Pd/e/JLkpydXTuW/t7g/2bKb+/pZrMaeqOpjkh5O8eTquJFcluXHqcvoYnRq7G5O8ZOp/TZJ3\ndPdXu/sTSY5lNufMuwWoqqdm9svwLUnS3Y909xdiLu0160meWFXrSZ6U5DMxl4br7g8k+fxpzbsx\nd872GpzmTGPU3X/V3Y9Ohx9McnDa39EceZy/0zjNWeZRkvxqkl9IsnUVPvNoxSwykD07yae2HB+f\n2tgF02MAVya5Ockzu/sz06nPJnnmtH+2MXqs9uNnaGdnfi2zv0z/fTq+OMkXtvwi3Ppz/dpYTOe/\nOPXf6dixM5clOZHkd2r2aOmbq+rCmEt7Rnc/kOT/JfmnzILYF5PcEnNpr9qNuXO212DnfiqzuybJ\nzsfo8fxOYw5VdU2SB7r7I6edMo9WjEU9VkBVPTnJnyS5rru/tPXc9C8h/m+DQarqlUke7O5bRtfC\nY1rP7FGR3+7uK5M8nNmjG19jLo01fa7hmszC87OSXJjk6qFFMZfdmDvm5+NXVW9K8miSt42uha+r\nqicl+cUk/3u3XtM8GmeRgeyBzJ5zPeXg1MYSVdV5mYWxt3X3u6fmf55uT2f6+uDUfrYxeqz2g2do\nZ34vTvKqqro/s8c7rkry65k9XrA+9dn6c/3aWEznn5rkoex87NiZ40mOd/fN0/GNmQU0c2nv+MEk\nn+juE939b0nendn8Mpf2pt2YO2d7DeZUVa9L8sokPza9GU92PkYPZefzkO09N7N/gPrI9B7iYJJb\nq+rbYx6tnEUGsn9I8rxppZ3zM/vw5nsXeH1OMz2H/ZYkd3X3r2w59d4kp1bWeW2SP9vS/hPT6jwv\nSvLF6Tb10SQvraqnTf8K/dIkR6dzX6qqF02v9RNbrsUcuvuN3X2wuy/NbE78bXf/WJL3J3n11O30\nMTo1dq+e+vfUfu20YtVlSZ6X2Qd0zbsF6O7PJvlUVT1/anpJkjtjLu0l/5TkRVX1pOlneGqMzKW9\naTfmztlegzlU1dWZPU7/qu7+1y2ndjRHpnm103nINrr7o939bd196fQe4nhmC7l9NubR6ukFrQ4y\nza9XZLbS371J3rTIa9vO+PP+3sxuLd+e5LZpe0Vmz2f/TZKPJ/nrJE+f+leS35rG56NJNrZc66cy\n++DusSQ/uaV9I8kd0/f8ZpIa/ec+V7ck35+vr7L4nMx+wR1L8sdJLpjanzAdH5vOP2fL979pGod7\nsmWFPvNuYeNzRZLNaT79aWYrVJlLe2hL8ktJ7p5+jn+Q2Spw5tL4cXl7Zp/r+7fM3jT+9G7MnbO9\nhm3uMTqW2eeNTr1/uGFL/x3NkcczD23bj9Fp5+/P11dZNI9WbDs1GAAAAOwyi3oAAAAMIpABAAAM\nIpABAAAMIpABAAAMIpABAAAMIpABAAAMIpABAAAM8h8vuraUpZUO8AAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 1080x360 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.figure(figsize=(15, 5))\n",
    "plt.imshow(final_states, aspect='auto')\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
